mllib/src/main/scala/org/apache/spark/ml/feature/MinMaxScaler.scala

```diff
@@ -117,11 +113,75 @@ class MinMaxScaler @Since("1.5.0") (@Since("1.5.0") override val uid: String)
   @Since("2.0.0")
   override def fit(dataset: Dataset[_]): MinMaxScalerModel = {
     transformSchema(dataset.schema, logging = true)
-    val input: RDD[OldVector] = dataset.select($(inputCol)).rdd.map {
-      case Row(v: Vector) => OldVectors.fromML(v)
+
+    val (mins, maxs, nnz, cnt) = dataset.select($(inputCol)).rdd.map {
+      row => row.getAs[Vector](0)
+    }.treeAggregate[(Array[Double], Array[Double], Array[Long], Long)](
+      (Array.emptyDoubleArray, Array.emptyDoubleArray, Array.emptyLongArray, 0L))(
+      seqOp = {
+        case ((min, max, nnz, cnt), vec) if cnt == 0 =>
+          val n = vec.size
+          val min_ = Array.fill[Double](n)(Double.MaxValue)
+          val max_ = Array.fill[Double](n)(Double.MinValue)
+          val nnz_ = Array.fill[Long](n)(0L)
+          vec.foreachActive {
+            case (i, v) if v != 0.0 =>
+              if (v < min_(i)) {
+                min_(i) = v
+              }
+              if (v > max_(i)) {
+                max_(i) = v
+              }
+              nnz_(i) = 1L
+            case _ =>
+          }
+          (min_, max_, nnz_, 1L)
+        case ((min, max, nnz, cnt), vec) =>
+          require(min.length == vec.size,
+            s"Dimensions mismatch when adding new sample: ${min.length} != ${vec.size}")
+          vec.foreachActive {
+            case (i, v) if v != 0.0 =>
```

Comment:
this looks the same as the code above, can you refactor to a separate function and call it from both places?